// proskomma-json-tools/dist/transforms/perf2x/wordCountActions.js
"use strict";

var wordCountActions = {
  startDocument: [{
    description: "Set up word object",
    test: () => true,
    action: _ref => {
      var {
        config,
        context,
        workspace,
        output
      } = _ref;
      workspace.words = {};
    }
  }],
  text: [{
    description: "Split strings and add words to word object",
    test: () => true,
    action: _ref2 => {
      var {
        config,
        context,
        workspace,
        output
      } = _ref2;
      for (var word of context.sequences[0].element.text.split(/[\s:;,.]+/).filter(w => w.length > 0)) {
        word = word.toLowerCase();
        if (word in workspace.words) {
          workspace.words[word] += 1;
        } else {
          workspace.words[word] = 1;
        }
      }
    }
  }],
  endDocument: [{
    description: "Sort words",
    test: () => true,
    action: _ref3 => {
      var {
        config,
        context,
        workspace,
        output
      } = _ref3;
      output.words = [...Object.entries(workspace.words)].sort((a, b) => b[1] - a[1]);
    }
  }]
};
module.exports = wordCountActions;
