// proskomma-json-tools/dist/render/perfToX/renderActions/calculateUsfmChapterPositions.js
"use strict";

var initialBlockRecord = ct => ({
  type: ct.sequences[0].block.type,
  subType: ct.sequences[0].block.subType,
  pos: ct.sequences[0].block.blockN,
  perfChapter: null
});
var calculateUsfmChapterPositionsActions = {
  startDocument: [{
    description: "Set up storage",
    test: () => true,
    action: _ref => {
      var {
        workspace,
        output
      } = _ref;
      workspace.blockRecords = [];
      output.report = {};
    }
  }],
  startParagraph: [{
    description: "Set up block record",
    test: () => true,
    action: _ref2 => {
      var {
        context,
        workspace,
        output
      } = _ref2;
      workspace.blockRecords.push(initialBlockRecord(context));
    }
  }],
  blockGraft: [{
    description: "Set up block record",
    test: () => true,
    action: _ref3 => {
      var {
        context,
        workspace,
        output
      } = _ref3;
      workspace.blockRecords.push(initialBlockRecord(context));
    }
  }],
  mark: [{
    description: "Add chapter number to block record",
    test: _ref4 => {
      var {
        context
      } = _ref4;
      return context.sequences[0].element.subType === "chapter";
    },
    action: _ref5 => {
      var {
        config,
        context,
        workspace,
        output
      } = _ref5;
      workspace.blockRecords[workspace.blockRecords.length - 1].perfChapter = context.sequences[0].element.atts["number"];
    }
  }],
  endDocument: [{
    description: "Populate report",
    test: () => true,
    action: _ref6 => {
      var {
        workspace,
        output
      } = _ref6;
      for (var [recordN, record] of Object.entries(workspace.blockRecords)) {
        if (!record.perfChapter) {
          continue;
        }
        var usfmChapterPos = recordN;
        var found = false;
        while (usfmChapterPos > 0 && !found) {
          if (workspace.blockRecords[usfmChapterPos - 1].type === 'paragraph' || workspace.blockRecords[usfmChapterPos - 1].subType === 'title') {
            found = true;
          } else {
            usfmChapterPos--;
          }
        }
        output.report[usfmChapterPos.toString()] = record.perfChapter;
      }
    }
  }]
};
module.exports = {
  calculateUsfmChapterPositionsActions
};
