// proskomma-json-tools/dist/render/alignment/transforms/mergeAlignment.js
"use strict";

var _PerfRenderFromJson = _interopRequireDefault(require("../../../PerfRenderFromJson"));
var _mergeActions = _interopRequireDefault(require("../../../mergeActions"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var {
  identityActions
} = require('../../perfToPerf/renderActions/identity');
var {
  mergeAlignmentActions
} = require('../renderActions/mergeAlignement');
var mergeAlignmentCode = function mergeAlignmentCode(_ref) {
  var {
    perf,
    verseWords: totalOccurrences,
    strippedAlignment
  } = _ref;
  var cl = new _PerfRenderFromJson.default({
    srcJson: perf,
    actions: (0, _mergeActions.default)([mergeAlignmentActions, identityActions])
  });
  var output = {};
  cl.renderDocument({
    docId: "",
    config: {
      totalOccurrences,
      strippedAlignment
    },
    output
  });
  return {
    perf: output.perf,
    unalignedWords: output.unalignedWords
  }; // identityActions currently put PERF directly in output
};

var mergeAlignment = {
  name: "mergeAlignment",
  type: "Transform",
  description: "PERF=>PERF adds report to verses",
  inputs: [{
    name: "perf",
    type: "json",
    source: ""
  }, {
    name: "strippedAlignment",
    type: "json",
    source: ""
  }, {
    name: "verseWords",
    type: "json",
    source: ""
  }],
  outputs: [{
    name: "perf",
    type: "json"
  }, {
    name: "unalignedWords",
    type: "json"
  }],
  code: mergeAlignmentCode
};
module.exports = {
  mergeAlignment
};
