// proskomma-json-tools/dist/render/alignment/transforms/stripAlignment.js
"use strict";

var _PerfRenderFromJson = _interopRequireDefault(require("../../../PerfRenderFromJson"));
var _mergeActions = _interopRequireDefault(require("../../../mergeActions"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var {
  identityActions
} = require('../../perfToPerf/renderActions/identity');
var {
  stripMarkupActions
} = require('../renderActions/stripAlignment');
var stripMarkupCode = function stripMarkupCode(_ref) {
  var {
    perf,
    verseWords
  } = _ref;
  var cl = new _PerfRenderFromJson.default({
    srcJson: perf,
    actions: (0, _mergeActions.default)([stripMarkupActions, identityActions])
  });
  var output = {};
  cl.renderDocument({
    docId: "",
    config: {
      verseWords
    },
    output
  });
  return {
    perf: output.perf,
    strippedAlignment: output.stripped,
    unalignedWords: output.unalignedWords
  };
};
var stripAlignment = {
  name: "stripAlignment",
  type: "Transform",
  description: "PERF=>PERF: Strips alignment markup",
  inputs: [{
    name: "perf",
    type: "json",
    source: ""
  }, {
    name: "verseWords",
    type: "json",
    source: ""
  }],
  outputs: [{
    name: "perf",
    type: "json"
  }, {
    name: "strippedAlignment",
    type: "json"
  }, {
    name: "unalignedWords",
    type: "json"
  }],
  code: stripMarkupCode
};
module.exports = {
  stripAlignment
};
