// proskomma-json-tools/dist/render/perfToX/transforms/calculateUsfmChapterPositions.js
"use strict";

var _PerfRenderFromJson = _interopRequireDefault(require("../../../PerfRenderFromJson"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var {
  calculateUsfmChapterPositionsActions
} = require('../renderActions/calculateUsfmChapterPositions');
var calculateUsfmChapterPositionsCode = function calculateUsfmChapterPositionsCode(_ref) {
  var {
    perf
  } = _ref;
  var cl = new _PerfRenderFromJson.default({
    srcJson: perf,
    actions: calculateUsfmChapterPositionsActions
  });
  var output = {};
  cl.renderDocument({
    docId: "",
    config: {
      maxLength: 60
    },
    output
  });
  return {
    report: output.report
  };
};
var calculateUsfmChapterPositions = {
  name: "calculateUsfmChapterPositions",
  type: "Transform",
  description: "PERF=>JSON: Generates positions for inserting chapter numbers into USFM",
  inputs: [{
    name: "perf",
    type: "json",
    source: ""
  }],
  outputs: [{
    name: "report",
    type: "json"
  }],
  code: calculateUsfmChapterPositionsCode
};
module.exports = {
  calculateUsfmChapterPositions
};
