// proskomma-json-tools/dist/render/alignment/renderActions/verseWords.js
"use strict";

var _xregexp = _interopRequireDefault(require("xregexp"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var verseWordsActions = {
  startDocument: [{
    description: "Set up storage",
    test: () => true,
    action: _ref => {
      var {
        workspace,
        output
      } = _ref;
      workspace.verseContent = [];
      workspace.chapter = null;
      workspace.verses = null;
      output.cv = {};
    }
  }],
  mark: [{
    description: "Update CV state",
    test: () => true,
    action: _ref2 => {
      var {
        context,
        workspace,
        output
      } = _ref2;
      var {
        element
      } = context.sequences[0];
      if (element.subType === 'chapter') {
        workspace.chapter = element.atts['number'];
        workspace.verses = 0;
        output.cv[workspace.chapter] = {};
        output.cv[workspace.chapter][workspace.verses] = {};
      } else if (element.subType === 'verses') {
        workspace.verses = element.atts['number'];
        output.cv[workspace.chapter][workspace.verses] = {};
      }
    }
  }],
  text: [{
    description: "Log occurrences",
    test: () => true,
    action: _ref3 => {
      var {
        context,
        workspace,
        output
      } = _ref3;
      var {
        chapter,
        verses
      } = workspace;
      var {
        text
      } = context.sequences[0].element;
      var re = (0, _xregexp.default)('([\\p{Letter}\\p{Number}\\p{Mark}\\u2060]{1,127})');
      var words = _xregexp.default.match(text, re, "all");
      for (var word of words) {
        var _output$cv$chapter$ve, _output$cv$chapter$ve2;
        (_output$cv$chapter$ve2 = (_output$cv$chapter$ve = output.cv[chapter][verses])[word]) !== null && _output$cv$chapter$ve2 !== void 0 ? _output$cv$chapter$ve2 : _output$cv$chapter$ve[word] = 0;
        output.cv[chapter][verses][word] += 1;
      }
    }
  }]
};
module.exports = {
  verseWordsActions
};
