// proskomma-tools/dist/pre-query/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.searchTermsRegex = exports.searchTermsClause = exports.searchTerms = exports.parseReferenceString = exports.default = exports.attTermsClause = void 0;

function _wrapRegExp() { _wrapRegExp = function _wrapRegExp(re, groups) { return new BabelRegExp(re, void 0, groups); }; var _super = RegExp.prototype, _groups = new WeakMap(); function BabelRegExp(re, flags, groups) { var _this = new RegExp(re, flags); return _groups.set(_this, groups || _groups.get(re)), _setPrototypeOf(_this, BabelRegExp.prototype); } function buildGroups(result, re) { var g = _groups.get(re); return Object.keys(g).reduce(function (groups, name) { return groups[name] = result[g[name]], groups; }, Object.create(null)); } return _inherits(BabelRegExp, RegExp), BabelRegExp.prototype.exec = function (str) { var result = _super.exec.call(this, str); return result && (result.groups = buildGroups(result, this)), result; }, BabelRegExp.prototype[Symbol.replace] = function (str, substitution) { if ("string" == typeof substitution) { var groups = _groups.get(this); return _super[Symbol.replace].call(this, str, substitution.replace(/\$<([^>]+)>/g, function (_, name) { return "$" + groups[name]; })); } if ("function" == typeof substitution) { var _this = this; return _super[Symbol.replace].call(this, str, function () { var args = arguments; return "object" != typeof args[args.length - 1] && (args = [].slice.call(args)).push(buildGroups(args, _this)), substitution.apply(this, args); }); } return _super[Symbol.replace].call(this, str, substitution); }, _wrapRegExp.apply(this, arguments); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var searchTerms = text => text.split(/ +/).map(term => term.trim()).filter(term => term.length > 0).filter(term => !term.includes(':'));

exports.searchTerms = searchTerms;

var searchTermsClause = text => searchTerms(text).map(st => "\"".concat(st.toLowerCase(), "\"")).join(', ');

exports.searchTermsClause = searchTermsClause;

var attTermsClause = text => text.split(/ +/).map(term => term.trim()).filter(term => term.length > 0).filter(term => term.includes(':')).map(term => term.split(':').slice(0, 2)).map(st => {
  var isMilestone = st[0].startsWith('x-');
  var attribute = isMilestone ? 'milestone' : 'spanWithAtts';
  var marker = isMilestone ? 'zaln' : 'w';
  return "\"\"\"attribute/".concat(attribute, "/").concat(marker, "/").concat(st[0], "/0/").concat(st[1], "\"\"\"");
}).join(', ');

exports.attTermsClause = attTermsClause;

var searchTermsRegex = text => {
  var _searchTerms = searchTerms(text);

  var regex = 'xxxxx';

  if (_searchTerms.length > 0) {
    regex = _searchTerms.map(st => "(".concat(st, ")")).join('|');
  }

  ;
  return regex;
};

exports.searchTermsRegex = searchTermsRegex;

var parseReferenceString = reference => {
  var response = {}; // 3JN 1:1-2 PSA 119:100 MAT 1-2

  var regex = /*#__PURE__*/_wrapRegExp(/([\d\w]\w{2}) ((\d+):?([\d-]*))/, {
    bookCode: 1,
    cv: 2,
    c: 3,
    v: 4
  });

  var {
    bookCode,
    cv,
    c,
    v
  } = reference.match(regex).groups || {};
  response.bookCode = bookCode;

  if (cv.includes(':') && !!v) {
    response.chapterVerses = cv;
  } else if (!!c && !v) {
    response.chapter = c;
  }

  ;
  return response;
};

exports.parseReferenceString = parseReferenceString;
var preQueries = {
  searchTerms,
  searchTermsClause,
  attTermsClause,
  searchTermsRegex,
  parseReferenceString
};
var _default = preQueries;
exports.default = _default;
