// proskomma-json-tools/dist/render/perfToPerf/transforms/justTheBible.js
"use strict";

var _PerfRenderFromJson = _interopRequireDefault(require("../../../PerfRenderFromJson"));
var _mergeActions = _interopRequireDefault(require("../../../mergeActions"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var {
  identityActions
} = require('../../perfToPerf/renderActions/identity');
var {
  justTheBibleActions
} = require('../../perfToPerf/renderActions/justTheBible');
var justTheBibleCode = function justTheBibleCode(_ref) {
  var {
    perf
  } = _ref;
  var cl = new _PerfRenderFromJson.default({
    srcJson: perf,
    actions: (0, _mergeActions.default)([justTheBibleActions, identityActions])
  });
  var output = {};
  cl.renderDocument({
    docId: "",
    config: {},
    output
  });
  return {
    perf: output.perf
  }; // identityActions currently put PERF directly in output
};

var justTheBible = {
  name: "justTheBible",
  type: "Transform",
  description: "PERF=>PERF: Strips most markup",
  documentation: "This transform removes milestones, wrappers and most marks. It has been used in several pipelines. It may also be stripping metaContent.",
  inputs: [{
    name: "perf",
    type: "json",
    source: ""
  }],
  outputs: [{
    name: "perf",
    type: "json"
  }],
  code: justTheBibleCode
};
module.exports = {
  justTheBible
};
